### ISSUE TITLE
Add an option to show the N lines before/after a filtered line for additional context

### ISSUE THREAD

**Issue by DavidAnson** [TextAnalysisTool/Releases#6]
This would be similar to how diff-ing tools show 5 lines before/after a diff. N should be configurable.

**Comment by JonnyOThan** [comment]
(Copying my comment from an issue I opened before noticing this one)

Suppose I have a lot of lines like:
A
A
A
B
A
A
A
B
A

The "B" line is essentially a marker, and what I'm really interested is in the "A" line immediately preceding the "B" line. But there are lots of "A" lines in the file that I don't care about.

I'm not entirely sure what the best way to specify something like this would be:

Maybe in the Edit Filter window there could be a bit of UI that says "include nearby line ranges" where you can enter some numbers like "-3,-1" to include the 3 lines immediately preceding the one that matches the filter.

Another possibility would be to allow for a compound filter made up of several single-line filters where each filter has to match a consecutive line. This might address some of the use cases in https://github.com/TextAnalysisTool/Releases/issues/30 as well.

**Comment by MartinSkold** [comment]
Hi @DavidAnson! I saw you created this issue a very long time ago

I been using this tool the last 5 years, and I believe that this is the only thing I am missing. 
Is it technically feasable to implement this? Or is it near impossible due to the architecture of the program?

Thanks again for the making of this tool. I have not seen any other program that is so simple, fast and have such a easy to use ux.

**Comment by DavidAnson** [comment]
Thank you! I think it's possible. I have been spending my personal project time elsewhere, though.

**Comment by vincec-msft** [comment]
My questions would be more about the UX than the technical possibility.
- Should the number of lines be adjustable while viewing or just during filtering (this makes a difference on large files where refiltering could be slow, but changing on the fly would be more complicated).
- Should the additional lines have a different style that shows they are not a match, but only provided for context?
- Should there be a different number of lines before/after?
- Should each filter have a different number of lines before/after?
- Should the additional context be easily toggleable?

Bottom line: How simple could we make it so that it's still useful?

**Comment by MartinSkold** [comment]
@vincec-msft A lot of great questions! 

I can tell my workflow, and then maybe you can tell me how you use the tool.

Right now, I try to create a filter to match a certain message, and then create another filter to match the lines around it as a workaround (very rarely this is not possible, and it can be quite cumbersome). 
Usually I know that before or after a certain message, there will be x lines of messages (usually the same messages appear before). The context is mostly static. 

For your question, my opinion:
- **Should the number of lines be adjustable while viewing or just during filtering (this makes a difference on large files where refiltering could be slow, but changing on the fly would be more complicated).**
    - _Unnecessarily_: Per filter settings would be easiest and best in my view. I have not seen any usecase to modify context size during inspection of a file. Then I would just toggle option "show only filtered lines", if I need to compare. Since it stays on the marked line, it is at least a way to get context. In my logs, there is usually x lines before/after. And if tje context varied, I would in that case just set a more number of lines as context. And if the order in the file is not always the same, I would assume that you create more filters. 

- **Should the additional lines have a different style that shows they are not a match, but only provided for context?**
    - _Unnecessarily_: To me, that would not matter. I like to color everything related to one message in one color. You can easily see which of the messages is the bottom one, since it would have the same color. Seems like it only add complexity in implementation and ux. 

- **Should there be a different number of lines before/after?**
    - _Yes please_! To me, the perfect implementation would be like grep's -A, -B, -C, that you can choose context before and after. For example 2 lines before and 4 lines after. 

- **Should each filter have a different number of lines before/after?**
    - I would say that it would be most useful if it configurable per filter, in the add filter menu. If I filter a message, I know what context I need. 

- **Should the additional context be easily toggleable?**
    - I can see that it would be convenient. To me it would be enough with a toggle in the add filter menu when you edit the filter, and the rows stay there when you untoggle it. And ofc a toggle in the menu would be nice. But nothing of it would be important to me. I am so happy with this tool compared to any alternatives. 


A implementation issue to figure out would be what happens when context overlaps a line from another filter. 
- Should it continue to add context, or stop there? 
- And in that case, what color?
- If both have context enabled, what color should the context get?
I don't know if you have a good answer or input for those questions.
